src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java

```diff
@@ -0,0 +1,1753 @@
+        boolean hasResultSet = statement.execute(query);
+        if (hasResultSet) {
+          statement.getResultSet().close();
+        }
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Return a query that can be used to check the validity of an existing database connection
+   * when the JDBC driver does not support JDBC 4. By default this returns {@code SELECT 1},
+   * but subclasses should override this when a different query should be used.
+   *
+   * @return the check connection query; may be null if the connection should not be queried
+   */
+  protected String checkConnectionQuery() {
+    return "SELECT 1";
+  }
+
+  protected JdbcDriverInfo jdbcDriverInfo() {
+    if (jdbcDriverInfo == null) {
+      try (Connection connection = getConnection()) {
+        jdbcDriverInfo = createJdbcDriverInfo(connection);
+      } catch (SQLException e) {
+        throw new ConnectException("Unable to get JDBC driver information", e);
+      }
+    }
+    return jdbcDriverInfo;
+  }
+
+  protected JdbcDriverInfo createJdbcDriverInfo(Connection connection) throws SQLException {
+    DatabaseMetaData metadata = connection.getMetaData();
+    return new JdbcDriverInfo(metadata.getJDBCMajorVersion(), metadata.getJDBCMinorVersion(),
+                              metadata.getDriverName(), metadata.getDatabaseProductName(),
+                              metadata.getDatabaseProductVersion()
+    );
+  }
+
+  /**
+   * Add or modify any connection properties based upon the {@link #config configuration}.
+   *
+   * <p>By default this method does nothing and returns the {@link Properties} object supplied as a
+   * parameter, but subclasses can override it to add/remove properties used to create new
+   *
+   * @param properties the properties that will be passed to the {@link DriverManager}'s {@link
+   *                   DriverManager#getConnection(String, Properties) getConnection(...) method};
+   *                   never null
+   * @return the updated connection properties, or {@code properties} if they are not modified or
+   *     should be returned; never null
+   */
+  protected Properties addConnectionProperties(Properties properties) {
+    return properties;
+  }
+
+  @Override
+  public PreparedStatement createPreparedStatement(
+      Connection db,
+      String query
+  ) throws SQLException {
+    PreparedStatement stmt = db.prepareStatement(query);
+    stmt.setFetchDirection(ResultSet.FETCH_FORWARD);
+    return stmt;
+  }
+
+  @Override
+  public TableId parseTableIdentifier(String fqn) {
+    List<String> parts = identifierRules().parseQualifiedIdentifier(fqn);
+    if (parts == null || parts.isEmpty()) {
+      throw new IllegalArgumentException("Invalid fully qualified name: '" + fqn + "'");
+    }
+    if (parts.size() == 1) {
+      return new TableId(null, null, parts.get(0));
+    }
+    if (parts.size() == 3) {
+      return new TableId(parts.get(0), parts.get(1), parts.get(2));
+    }
+    assert parts.size() >= 2;
```

Comment:
this is always true

Reply:
Correct, which is why I used an assertion. It's a great way to clarify the logic, but since asserts can be turned off there's no harm in stating them.